src/dataflow/src/render/envelope_none.rs

```diff
@@ -0,0 +1,122 @@
+// by the Apache License, Version 2.0.
+
+use persist::client::{StreamReadHandle, StreamWriteHandle};
+use timely::dataflow::operators::{Concat, Map, OkErr};
+use timely::dataflow::{Scope, Stream};
+use timely::progress::Antichain;
+
+use dataflow_types::{DataflowError, DecodeError, SourceError, SourceErrorDetails};
+use persist::operators::replay::Replay;
+use persist::operators::stream::{Persist, RetractUnsealed};
+use persist_types::Codec;
+use repr::{Diff, Row, Timestamp};
+
+use crate::operator::StreamExt;
+
+/// Persist configuration for `ENVELOPE NONE` sources.
+#[derive(Debug, Clone)]
+pub struct PersistentEnvelopeNoneConfig<V: Codec> {
+    /// The timestamp up to which which data should be read when restoring.
+    pub upper_seal_ts: u64,
+
+    /// [`StreamReadHandle`] for the collection that we should persist to.
+    pub read_handle: StreamReadHandle<V, ()>,
+
+    /// [`StreamWriteHandle`] for the collection that we should persist to.
+    pub write_handle: StreamWriteHandle<V, ()>,
+}
+
+impl<V: Codec> PersistentEnvelopeNoneConfig<V> {
+    /// Creates a new [`PersistentEnvelopeNoneConfig`] from the given parts.
+    pub fn new(
+        upper_seal_ts: u64,
+        read_handle: StreamReadHandle<V, ()>,
+        write_handle: StreamWriteHandle<V, ()>,
+    ) -> Self {
+        PersistentEnvelopeNoneConfig {
+            upper_seal_ts,
+            read_handle,
+            write_handle,
+        }
+    }
+}
+
+/// Persists the given input stream, passes through the data it carries and replays previously
+/// persisted updates when starting up.
+///
+/// This will filter out and retract replayed updates that are not beyond the `upper_seal_ts` given
+/// in `persist_config`.
+///
+pub(crate) fn persist_and_replay<G>(
+    source_name: &str,
+    stream: &Stream<G, Result<Row, DecodeError>>,
+    as_of_frontier: &Antichain<Timestamp>,
+    persist_config: PersistentEnvelopeNoneConfig<Result<Row, DecodeError>>,
+) -> (
+    Stream<G, (Result<Row, DecodeError>, Timestamp, Diff)>,
+    Stream<G, (DataflowError, Timestamp, Diff)>,
+)
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let scope = stream.scope();
+
+    let (restored_oks, restored_errs) = {
+        let snapshot = persist_config.read_handle.snapshot();
+
+        let (restored_oks, restored_errs) =
+            scope.replay(snapshot, as_of_frontier).ok_err(split_ok_err);
+
+        let (restored_oks, retract_errs) = restored_oks.retract_unsealed(
+            source_name,
+            persist_config.write_handle.clone(),
+            persist_config.upper_seal_ts,
+        );
+
+        let combined_errs = restored_errs.concat(&retract_errs);
+
+        (restored_oks, combined_errs)
+    };
+
+    // Persist can only deal with (key, value) streams, so promote value to key.
+    let flattened_stream = stream
+        .map(|v| (v, ()))
+        .pass_through(format!("pass-through({})", source_name).as_str());
```

Comment:
can we move the pass_through up before this operator? it's subjective of course, but that feels to me to be more "source decoding" logic than the rest of what's in here

Reply:
:ok_hand: